src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,598 @@
+@param[in] ref     Reference level for calibration
+@note   These parameters must be set before the plugin can calculate the level
+        They are determined by the plugin configuration
+/*****************************************************************************/
+void P145_data_struct::setSensorData(int stype, bool comp, bool cal, float load, float zero, float ref)
+{
+  /* Each MQ-xxx sensor comes with its own set cof constants */
+  /* Copy the correct set from program meory space           */
+  if ((stype != sensorType) && (stype < nbrOfTypes) && (stype >= 0))
+  {
+    memcpy_P(&sensordef, &sensorData[stype], sizeof(struct P145_SENSORDEF));
+    algorithm = sensordef.alg;
+    sensorType = stype;   // Selected sensor type, index in sensor data table
+  }
+  compensation = comp;    // Compensation selection flag
+  calibration = cal;      // Calibration selection flag
+  rload = load;           // Rload, load resistor [Ohm]
+  rzero = zero;           // R0, reference resistance [Ohm]
+  refLevel = ref;         // Reference level for calibration [ppm]
+}
+
+/*****************************************************************************/
+/*!
+@brief  Return the current/latest value acquired from the sensor 
+@param[in] temperature  Temperature for compensation
+@param[in] humidity     Humidity for compensation
+@return The calculated gas concentration [ppm or mg/l]
+@note   This is the entry point to start collecting the oversampling data,
+        convert the analog value to Rsensor and convert Rsensor to the gas
+        concentration level. It also performs autocalibration if selected.
+        Oversampling data will be reset as a side effec.
+        This function shall be called at regular interval for proper performance.
+*/
+/*****************************************************************************/
+float P145_data_struct::readValue(float temperature, float humidity)
+{
+    float ain = getAnalogValue();           // Analog measured and filtered data 
+    float rSensor = getResistance(ain);     // Convert to sensor resistance Rs
+    float rCal    = 0.0;                    // Potential Rzero for calibration
+    float value = 0.0;                      // Return value
+
+    // Check if temperature/Humidity compensation is selected
+    if (compensation)
+    {
+        value = getCorrectedPPM(rSensor, temperature, humidity);
+        rCal = getCorrectedRZero(rSensor, temperature, humidity);
+    }
+    else
+    {
+        value = getPPM(rSensor);
+        rCal = getRZero(rSensor);
+    }
+
+    // Perform calibration if functionality is enabled
+    if (calibration && (rCal < 1.0e6)) 
+    { 
+        calibrate(rCal);
+    }
+
+    if (loglevelActiveFor(LOG_LEVEL_INFO))
+    {
+      addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: level= "), value));         // Calculated sensor value
+#ifdef P145_DEBUG
+      addLog(LOG_LEVEL_INFO, concat(concat(F("MQ-xx: Sensor type= "), sensorType), concat(F(": "), sensordef.name))); 
```

Comment:
Please wrap the xxx.name and xxx.gas types into something like this:  `String(sensordef.name)` when used along with `concat`
The reason is that `concat` is a template function.
And since `name` is of type `char[8]`, the compiler will create a new instance of each call like this.
Right now `gas` and `name` are of the same length, but when calling concat with a `char[9]` it will generate new code for the same function.